realm-transformer/src/main/kotlin/io/realm/transformer/IncrementalBuild.kt

```diff
@@ -0,0 +1,134 @@
+        inputs.forEach {
+            // Files in directories are files we most likely want to transform
+             it.directoryInputs.forEach {
+                val dirPath: String = it.file.absolutePath
+                it.changedFiles.entries.forEach {
+                    if (it.value == Status.NOTCHANGED || it.value == Status.REMOVED) {
+                        return
+                    }
+                    val filePath: String = it.key.absolutePath
+                    if (filePath.endsWith(SdkConstants.DOT_CLASS)) {
+                        val className = filePath
+                                .substring(dirPath.length + 1, filePath.length - SdkConstants.DOT_CLASS.length)
+                                .replace(File.separatorChar, '.')
+                        directoryFiles.add(className)
+                    }
+                }
+            }
+
+            // Files in Jars are always treated as referenced input. They should already have been
+            // modified by the transformer in the project that built the jar.
+            it.jarInputs.forEach {
+                if (it.status == Status.REMOVED) {
+                    return
+                }
+
+                val jarFile = JarFile(it.file)
+                jarFile.entries()
+                        .toList()
+                        .filter {
+                            !it.isDirectory && it.name.endsWith(SdkConstants.DOT_CLASS)
+                        }
+                        .forEach {
+                            val path: String = it.name
+                            // The jar might not using File.separatorChar as the path separator. So we just replace both `\` and
+                            // `/`. It depends on how the jar file was created.
+                            // See http://stackoverflow.com/questions/13846000/file-separators-of-path-name-of-zipentry
+                            val className: String = path
+                                    .substring(0, path.length - SdkConstants.DOT_CLASS.length)
+                                    .replace('/', '.')
+                                    .replace('\\', '.')
+                            referencedFiles.add(className)
+                        }
+                jarFile.close() // Crash transformer if this fails
+            }
+        }
+    }
+
+    override fun findModelClasses(classNames: Set<String>): Collection<CtClass> {
+        val realmObjectProxyInterface: CtClass = classPool.get("io.realm.internal.RealmObjectProxy")
+        // For incremental builds we need to determine if a class is a model class file
+        // based on information in the file itself. This require checks that are only
+        // possible once we loaded the CtClass from the ClassPool and is slower
+        // than the approach used when doing full builds.
+        return classNames
+                // Map strings to CtClass'es.
+                .map { classPool.getCtClass(it) }
```

Comment:
why not using the same heuristic as full build where you check first if the file ends with the `RealmProxy` then load the `CtClass` to check the annotation?

Reply:
Because in a incremental build, the RealmProxy classes are only part of the build if the Realm model class is modified, so for incremental builds we need to be able to detect it from the model class itself, which is a bit more expensive as it requires looking at interfaces and superclasses.